src/Backups/BackupIO_Disk.cpp

```diff
@@ -8,138 +8,122 @@
 {
-    auto file_path = path / file_name;
+    auto file_path = root_path / file_name;
     disk->createDirectories(file_path.parent_path());
-    return disk->writeFile(file_path);
+    return disk->writeFile(file_path, write_buffer_size, WriteMode::Rewrite, write_settings);
 }
 
 void BackupWriterDisk::removeFile(const String & file_name)
 {
-    disk->removeFileIfExists(path / file_name);
-    if (disk->isDirectory(path) && disk->isDirectoryEmpty(path))
-        disk->removeDirectory(path);
+    disk->removeFileIfExists(root_path / file_name);
+    if (disk->isDirectory(root_path) && disk->isDirectoryEmpty(root_path))
+        disk->removeDirectory(root_path);
 }
 
 void BackupWriterDisk::removeFiles(const Strings & file_names)
 {
     for (const auto & file_name : file_names)
-        disk->removeFileIfExists(path / file_name);
-    if (disk->isDirectory(path) && disk->isDirectoryEmpty(path))
-        disk->removeDirectory(path);
-}
-
-DataSourceDescription BackupWriterDisk::getDataSourceDescription() const
-{
-    return disk->getDataSourceDescription();
-}
-
-DataSourceDescription BackupReaderDisk::getDataSourceDescription() const
-{
-    return disk->getDataSourceDescription();
+        disk->removeFileIfExists(root_path / file_name);
+    if (disk->isDirectory(root_path) && disk->isDirectoryEmpty(root_path))
+        disk->removeDirectory(root_path);
 }
 
-bool BackupWriterDisk::supportNativeCopy(DataSourceDescription data_source_description) const
+void BackupWriterDisk::copyFileFromDisk(const String & path_in_backup, DiskPtr src_disk, const String & src_path,
+                                        bool copy_encrypted, UInt64 start_pos, UInt64 length)
 {
-    return data_source_description == disk->getDataSourceDescription();
-}
-
-void BackupWriterDisk::copyFileNative(DiskPtr src_disk, const String & src_file_name, UInt64 src_offset, UInt64 src_size, const String & dest_file_name)
-{
-    if (!src_disk)
-        throw Exception(ErrorCodes::LOGICAL_ERROR, "Cannot natively copy data to disk without source disk");
-
-    if (has_throttling || (src_offset != 0) || (src_size != src_disk->getFileSize(src_file_name)))
+    /// Use IDisk::copyFile() as a more optimal way to copy a file if it's possible.
+    bool has_throttling = src_disk->isRemote() ? static_cast<bool>(read_settings.remote_throttler) : static_cast<bool>(read_settings.local_throttler);
+    if (!has_throttling && !start_pos && !copy_encrypted)
     {
-        auto create_read_buffer = [this, src_disk, src_file_name] { return src_disk->readFile(src_file_name, read_settings); };
-        copyDataToFile(create_read_buffer, src_offset, src_size, dest_file_name);
-        return;
+        auto source_data_source_description = src_disk->getDataSourceDescription();
+        if (source_data_source_description.sameKind(data_source_description) && !source_data_source_description.is_encrypted
+            && (length == src_disk->getFileSize(src_path)))
```

Comment:
Do we need to check length here if we already checked that `!start_pos == true`? I think in both cases it means that we need to copy the whole file.

Reply:
Why? It could be the first half of the file, for example.